### PR TITLE
Add caching for getUserIP and getSecureForwardedHeaderTrustedParts

### PR DESCRIPTION
**Pull request by rhertogh**
| Q             | A
| ------------- | ---
| Is bugfix?    | ❌
| New feature?  | ❌
| Breaks BC?    | ❌
| Fixed issues  | -

Many function in the `yii\web\Request` are cached in private properties. This PR adds the same functionality for the `getUserIP`  and `getSecureForwardedHeaderTrustedParts` functions.
This greatly improves performance when a lot of `trustedHosts` are set (e.g. when using CloudFlare or AWS CloudFront).

Before:
image

After:
image

### CODE REVIEW

File: framework/web/Request.php [hunk 1]
```diff
@@ -1220,15 +1220,21 @@ protected function getUserIpFromIpHeaders()
...
      * The IP is determined using headers and / or `$_SERVER` variables.
      * @return string|null user IP address, null if not available
      */
     public function getUserIP()
     {
-        $ip = $this->getUserIpFromIpHeaders();
-        return $ip === null ? $this->getRemoteIP() : $ip;
+        if ($this->_ip === null) {
+            $this->_ip = $this->getUserIpFromIpHeaders();
+            $this->_ip = $this->_ip === null ? $this->getRemoteIP() : $this->_ip;
```
Comment (bizley) [inline framework/web/Request.php:1234]:
```suggestion
            if ($this->_ip === null) {
                $this->_ip = $this->getRemoteIP();
            }
```

**Review by bizley — APPROVED** [review]

**Review by samdark — APPROVED** [review]
👍 Need a line for CHANGELOG and tests passing.

**Comment by bizley** [comment]
Looks like tests require updating.

**Comment by rhertogh** [comment]
@samdark The changlog line has been added.
@bizley The tests are updated, but there seems to be a problem with the test runners on Github.

**Comment by samdark** [comment]
Thank you!

**Pull request merged**